File: main/services/computer-use/process.ts
import { spawn, type ChildProcess } from "node:child_process";
import type { CuaDriverInvocation } from "./contract.js";
import { CuaDriverError, isAbortError } from "./contract.js";

const MAX_STDOUT_BYTES = 256 * 1024;
const MAX_STDERR_BYTES = 16 * 1024;
const TERMINATE_GRACE_MS = 500;
const KILL_GRACE_MS = 1_000;

export interface BoundedProcessResult {
  stdout: string;
  stderr: string;
}

function isChildReaped(child: ChildProcess): boolean {
  return child.exitCode !== null || child.signalCode !== null;
}

function waitForChildReaped(child: ChildProcess, timeoutMs: number): Promise<boolean> {
  if (isChildReaped(child)) return Promise.resolve(true);
  return new Promise((resolve) => {
    let done = false;
    const finish = (reaped: boolean) => {
      if (done) return;
      done = true;
      clearTimeout(timer);
      child.removeListener("exit", exited);
      child.removeListener("close", exited);
      resolve(reaped);
    };
    const exited = () => finish(true);
    const timer = setTimeout(() => finish(isChildReaped(child)), timeoutMs);
    timer.unref();
    child.once("exit", exited);
    child.once("close", exited);
  });
}

/**
 * Terminate a bridge or test broker that intentionally owns no same-process-
 * group descendants. Once the direct child is known reaped, never signal its
 * PID/PGID again: it may already have been reused by the OS.
 */
export async function terminateDirectChild(
  child: ChildProcess,
  options: { terminateGraceMs?: number; killGraceMs?: number } = {},
): Promise<void> {
  const terminateGraceMs = options.terminateGraceMs ?? TERMINATE_GRACE_MS;
  const killGraceMs = options.killGraceMs ?? KILL_GRACE_MS;
  if (!child.pid || isChildReaped(child)) return;
  let signalled = false;
  try {
    signalled = child.kill("SIGTERM");
  } catch {
    return;
  }
  if (!signalled || (await waitForChildReaped(child, terminateGraceMs))) return;
  // Recheck after the timer to close the exit-vs-escalation race.
  if (isChildReaped(child)) return;
  try {
    child.kill("SIGKILL");
  } catch {
    return;
  }
  child.stdin?.destroy();
  child.stdout?.destroy();
  child.stderr?.destroy();
  await waitForChildReaped(child, killGraceMs);
}

function destroyChildStreams(child: ChildProcess): void {
  child.stdin?.destroy();
  child.stdout?.destroy();
  child.stderr?.destroy();
}

export async function runCuaDriverCommand(
  invocation: CuaDriverInvocation,
  args: string[],
  options: {
    env: Record<string, string>;
    signal?: AbortSignal;
    timeoutMs?: number;
  },
): Promise<BoundedProcessResult> {
  if (options.signal?.aborted)
    throw new CuaDriverError("cancelled", "cua-driver request was cancelled.");
  return new Promise<BoundedProcessResult>((resolve, reject) => {
    const child = spawn(invocation.command, [...(invocation.prefixArgs ?? []), ...args], {
      // These are fixed platform utilities (`open`, `codesign`, and `plutil`).
      // Keep an exact ChildProcess handle instead of creating a process group:
      // a delayed negative-PID signal could target an unrelated reused PGID.
      detached: false,
      env: options.env,
      shell: false,
      stdio: ["ignore", "pipe", "pipe"],
      windowsHide: true,
    });
    const stdout: Buffer[] = [];
    const stderr: Buffer[] = [];
    let stdoutBytes = 0;
    let stderrBytes = 0;
    let settled = false;
    let terminationError: CuaDriverError | null = null;
    let timeoutTimer: NodeJS.Timeout | undefined;
    let terminationPromise: Promise<void> | null = null;

    const finish = (error?: Error) => {
      if (settled) return;
      settled = true;
      clearTimeout(timeoutTimer);
      options.signal?.removeEventListener("abort", abort);
      if (error) reject(error);
      else
        resolve({
          stdout: Buffer.concat(stdout).toString("utf8"),
          stderr: Buffer.concat(stderr).toString("utf8"),
        });
    };
    const terminate = (error: CuaDriverError) => {
      if (terminationError) return;
      terminationError = error;
      terminationPromise = terminateDirectChild(child).then(
        () => {
          destroyChildStreams(child);
          finish(error);
        },
        () => {
          destroyChildStreams(child);
          finish(error);
        },
      );
    };
    const abort = () =>
      terminate(new CuaDriverError("cancelled", "cua-driver request was cancelled."));
    timeoutTimer = setTimeout(
      () => terminate(new CuaDriverError("timeout", "cua-driver did not respond in time.", true)),
      options.timeoutMs ?? 5_000,
    );

    options.signal?.addEventListener("abort", abort, { once: true });
    if (options.signal?.aborted) abort();
    child.on("error", (error) => {
      if (terminationError) return;
      const code =
        (error as NodeJS.ErrnoException).code === "ENOENT" ? "driver_missing" : "spawn_failed";
      finish(new CuaDriverError(code, "The pinned cua-driver helper is unavailable."));
    });
    child.stdout.on("data", (chunk: Buffer) => {
      if (terminationError) return;
      stdoutBytes += chunk.byteLength;
      if (stdoutBytes > MAX_STDOUT_BYTES) {
        terminate(new CuaDriverError("output_too_large", "cua-driver returned too much output."));
      } else {
        stdout.push(chunk);
      }
    });
    child.stderr.on("data", (chunk: Buffer) => {
      if (terminationError) return;
      stderrBytes += chunk.byteLength;
      if (stderrBytes > MAX_STDERR_BYTES) {
        terminate(
          new CuaDriverError("output_too_large", "cua-driver returned too much diagnostic output."),
        );
      } else {
        stderr.push(chunk);
      }
    });
    child.on("close", (code, signal) => {
      if (settled) return;
      // Cancellation settles only after the exact direct child has been
      // supervised. Descendants are never inferred from a reusable PID/PGID.
      if (terminationError) void terminationPromise;
      else if (code === 0) finish();
      else {
        const diagnostic = Buffer.concat(stderr).toString("utf8").trim().slice(0, 600);
        finish(
          new CuaDriverError(
            "command_failed",
            `cua-driver exited ${signal ? `after ${signal}` : `with code ${String(code)}`}${diagnostic ? `: ${diagnostic}` : "."}`,
          ),
        );
      }
    });
  }).catch((error) => {
    if (error instanceof CuaDriverError || isAbortError(error)) throw error;
    throw new CuaDriverError("command_failed", "cua-driver command failed.");
  });
}

File: rust/aiden-computer-use/src/process.rs
//! Bounded subprocess execution and child termination (port of
//! `main/services/computer-use/process.ts`).

use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;
use std::time::Duration;

use tokio::process::Child;
use tokio_util::sync::CancellationToken;

use crate::contract::CuaDriverError;

const MAX_STDOUT_BYTES: u64 = 256 * 1024;
const MAX_STDERR_BYTES: u64 = 16 * 1024;
const TERMINATE_GRACE_MS: u64 = 500;
const KILL_GRACE_MS: u64 = 1_000;

/// A fixed platform utility invocation (`open`, `codesign`, `plutil`).
#[derive(Debug, Clone)]
pub struct CuaDriverCommandInvocation {
    pub command: PathBuf,
    pub prefix_args: Vec<String>,
}

impl CuaDriverCommandInvocation {
    pub fn new(command: impl Into<PathBuf>) -> Self {
        Self {
            command: command.into(),
            prefix_args: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Default)]
pub struct BoundedProcessResult {
    pub stdout: String,
    pub stderr: String,
}

/// Replace the inherited process environment with the caller's explicit
/// allowlist. Calling `envs` alone only overrides matching keys and otherwise
/// inherits every provider token, proxy credential, and loader hook owned by
/// the Aiden process.
pub(crate) fn apply_allowlisted_environment(
    command: &mut tokio::process::Command,
    environment: &HashMap<String, String>,
) {
    command.env_clear().envs(environment);
}

/// Wait until the child is reaped or the timeout elapses.
async fn wait_for_child_reaped(child: &mut Child, timeout_ms: u64) -> bool {
    tokio::time::timeout(Duration::from_millis(timeout_ms), child.wait())
        .await
        .is_ok()
}

/// Terminate a bridge or test broker that intentionally owns no same-process-
/// group descendants. Once the direct child is known reaped, never signal its
/// PID again: it may already have been reused by the OS.
#[cfg(unix)]
pub async fn terminate_direct_child(
    child: &mut Child,
    terminate_grace_ms: Option<u64>,
    kill_grace_ms: Option<u64>,
) {
    let terminate_grace_ms = terminate_grace_ms.unwrap_or(TERMINATE_GRACE_MS);
    let kill_grace_ms = kill_grace_ms.unwrap_or(KILL_GRACE_MS);
    let Some(pid) = child.id() else {
        return;
    };
    // SIGTERM first; escalate to SIGKILL after the grace period.
    unsafe {
        libc::kill(pid as libc::pid_t, libc::SIGTERM);
    }
    if wait_for_child_reaped(child, terminate_grace_ms).await {
        return;
    }
    let _ = child.start_kill();
    wait_for_child_reaped(child, kill_grace_ms).await;
}

/// A shared handle to a spawned bridge/broker child. The child is terminated
/// (SIGTERM → grace → SIGKILL) when `terminate` is called or when the last
/// handle is dropped, so a dropped session can never orphan the bridge.
pub struct ChildHandle {
    inner: std::sync::Mutex<Option<Child>>,
}

impl ChildHandle {
    pub fn new(child: Child) -> Arc<Self> {
        Arc::new(Self {
            inner: std::sync::Mutex::new(Some(child)),
        })
    }

    pub fn pid(&self) -> u32 {
        self.inner
            .lock()
            .unwrap()
            .as_ref()
            .and_then(|child| child.id())
            .unwrap_or(0)
    }

    pub fn terminate(&self) {
        let Some(child) = self.inner.lock().unwrap().take() else {
            return;
        };
        tokio::spawn(async move {
            let mut child = child;
            terminate_direct_child(&mut child, None, None).await;
        });
    }
}

impl Drop for ChildHandle {
    fn drop(&mut self) {
        if let Some(mut child) = self.inner.lock().unwrap().take() {
            let _ = child.start_kill();
        }
    }
}

/// Run a fixed platform utility with bounded output and an optional timeout,
/// aborting the child on cancellation (port of `runCuaDriverCommand`).
pub async fn run_command(
    invocation: &CuaDriverCommandInvocation,
    args: &[String],
    env: &HashMap<String, String>,
    signal: Option<&CancellationToken>,
    timeout_ms: u64,
) -> Result<BoundedProcessResult, CuaDriverError> {
    if signal.is_some_and(|token| token.is_cancelled()) {
        return Err(CuaDriverError::cancelled(
            "cua-driver request was cancelled.",
        ));
    }
    use std::process::Stdio;
    let mut command = tokio::process::Command::new(&invocation.command);
    command
        .args(invocation.prefix_args.iter())
        .args(args)
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    apply_allowlisted_environment(&mut command, env);
    let mut child = match command.spawn() {
        Ok(child) => child,
        Err(error) => {
            let code = if error.kind() == std::io::ErrorKind::NotFound {
                "driver_missing"
            } else {
                "spawn_failed"
            };
            return Err(CuaDriverError::new(
                code,
                "The pinned cua-driver helper is unavailable.",
            ));
        }
    };

    let stdout = child.stdout.take();
    let stderr = child.stderr.take();
    let stdout_task = stdout
        .map(|stream| tokio::spawn(async move { drain_bounded(stream, MAX_STDOUT_BYTES).await }));
    let stderr_task = stderr
        .map(|stream| tokio::spawn(async move { drain_bounded(stream, MAX_STDERR_BYTES).await }));

    let mut wait = Box::pin(child.wait());
    let deadline = std::time::Instant::now() + Duration::from_millis(timeout_ms);
    let mut stop: Option<CuaDriverError> = None;
    let mut wait_result: Option<Result<std::process::ExitStatus, std::io::Error>> = None;
    loop {
        if signal.is_some_and(|token| token.is_cancelled()) {
            stop = Some(CuaDriverError::cancelled(
                "cua-driver request was cancelled.",
            ));
            break;
        }
        let now = std::time::Instant::now();
        if now >= deadline {
            stop = Some(CuaDriverError::retryable(
                "timeout",
                "cua-driver did not respond in time.",
            ));
            break;
        }
        match tokio::time::timeout(deadline - now, &mut wait).await {
            Ok(result) => {
                wait_result = Some(result);
                break;
            }
            Err(_) => {
                // Deadline elapsed between checks; the loop re-evaluates.
            }
        }
    }
    drop(wait);
    if let Some(error) = stop {
        terminate_direct_child(&mut child, None, None).await;
        return Err(error);
    }
    let status = match wait_result.expect("wait settled") {
        Ok(status) => status,
        Err(_) => {
            return Err(CuaDriverError::new(
                "command_failed",
                "cua-driver command failed.",
            ))
        }
    };
    let stdout = match stdout_task {
        Some(task) => task.await.unwrap_or(Ok(String::new())).unwrap_or_default(),
        None => String::new(),
    };
    let stderr = match stderr_task {
        Some(task) => task.await.unwrap_or(Ok(String::new())).unwrap_or_default(),
        None => String::new(),
    };
    if status.success() {
        return Ok(BoundedProcessResult { stdout, stderr });
    }
    let diagnostic = stderr.trim().chars().take(600).collect::<String>();
    let message = match status.code() {
        Some(code) => format!(
            "cua-driver exited with code {code}{}",
            if diagnostic.is_empty() {
                ".".to_string()
            } else {
                format!(": {diagnostic}")
            }
        ),
        None => "cua-driver was terminated by a signal.".to_string(),
    };
    Err(CuaDriverError::new("command_failed", message))
}

async fn drain_bounded<R>(mut stream: R, maximum: u64) -> Result<String, CuaDriverError>
where
    R: tokio::io::AsyncRead + Unpin,
{
    use tokio::io::AsyncReadExt;
    let mut output = String::new();
    let mut bytes: u64 = 0;
    let mut buffer = [0_u8; 64 * 1024];
    loop {
        match stream.read(&mut buffer).await {
            Ok(0) | Err(_) => break,
            Ok(count) => {
                bytes += count as u64;
                if bytes > maximum {
                    return Err(CuaDriverError::new(
                        "output_too_large",
                        "cua-driver returned too much output.",
                    ));
                }
                output.push_str(&String::from_utf8_lossy(&buffer[..count]));
            }
        }
    }
    Ok(output)
}

#[cfg(test)]
mod tests {
    use super::*;

    const CHILD_MARKER: &str = "AIDEN_CUA_ENV_CLEAR_CHILD";
    const FORBIDDEN_PARENT_KEYS: &[&str] = &[
        "OPENAI_API_KEY",
        "CODEX_OAUTH_TOKEN",
        "HTTPS_PROXY",
        "DYLD_LIBRARY_PATH",
        "NODE_OPTIONS",
    ];

    #[tokio::test]
    async fn spawned_commands_receive_only_the_explicit_allowlist() {
        if std::env::var_os(CHILD_MARKER).is_none() {
            let current_executable = std::env::current_exe().expect("current test executable");
            let mut child = std::process::Command::new(current_executable);
            child
                .arg("--exact")
                .arg("process::tests::spawned_commands_receive_only_the_explicit_allowlist")
                .arg("--nocapture")
                .env_clear()
                .env(CHILD_MARKER, "1");
            for key in FORBIDDEN_PARENT_KEYS {
                child.env(key, format!("parent-secret-{key}"));
            }
            let output = child.output().expect("nested test process starts");
            assert!(
                output.status.success(),
                "nested environment test failed:\nstdout:\n{}\nstderr:\n{}",
                String::from_utf8_lossy(&output.stdout),
                String::from_utf8_lossy(&output.stderr)
            );
            return;
        }

        let environment = HashMap::from([
            ("PATH".to_string(), "/usr/bin:/bin".to_string()),
            ("HOME".to_string(), "/safe/aiden-home".to_string()),
            ("LANG".to_string(), "en_US.UTF-8".to_string()),
            (
                "CUA_DRIVER_HOST_BUNDLE_ID".to_string(),
                "com.aiden.safe-host".to_string(),
            ),
        ]);
        let result = run_command(
            &CuaDriverCommandInvocation::new("/usr/bin/env"),
            &[],
            &environment,
            None,
            2_000,
        )
        .await
        .expect("environment probe succeeds");
        let child_environment: HashMap<&str, &str> = result
            .stdout
            .lines()
            .filter_map(|line| line.split_once('='))
            .collect();

        assert_eq!(child_environment.get("PATH"), Some(&"/usr/bin:/bin"));
        assert_eq!(child_environment.get("HOME"), Some(&"/safe/aiden-home"));
        assert_eq!(child_environment.get("LANG"), Some(&"en_US.UTF-8"));
        assert_eq!(
            child_environment.get("CUA_DRIVER_HOST_BUNDLE_ID"),
            Some(&"com.aiden.safe-host")
        );
        assert!(!child_environment.contains_key(CHILD_MARKER));
        for key in FORBIDDEN_PARENT_KEYS {
            assert!(
                !child_environment.contains_key(key),
                "parent-only {key} reached the helper"
            );
        }
    }
}
